Importing libraries

In [64]:
import pandas as pd
import numpy as np
import os

In [65]:
path = r'C:\Users\Timbr\OneDrive\Desktop\Career Foundry\Data Immersion\Achievement 4\Instacart Basket Analysis - Achievement 4 -  Tim Brandt'

In [66]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'orders.csv'), index_col = False)

In [67]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

# Step 1 - Consistency Checks

In [68]:
# Create test dataframe:
df_test = pd.DataFrame()

In [69]:
# Create a mixed column type:
df_test['mix'] = ['a', 'b', 1, True]

In [70]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [71]:
# Check for mixed data types:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  mixed


In [72]:
# Changing 'mix' to string type:
df_test['mix'] = df_test['mix'].astype('str')

In [73]:
# Re-check for mixed data types:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  consistent


Missing Values

In [74]:
# Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [75]:
# Missing Values list: 
df_nanprods = df_prods[df_prods.product_name.isnull() == True]

In [80]:
df_nanprods

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [86]:
# Checking the number of rows in overall product dataframe
df_prods.shape

(49693, 5)

In [87]:
# Create new dataframe for non missing value
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [88]:
# Checking the number of rows in non missing value dataframe
df_prods_clean.shape

(49677, 5)

Duplicates

In [89]:
# Create duplicates df:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [90]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [92]:
# creating new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [93]:
# Checking the number of rows of clean data
df_prods_clean_no_dups.shape

(49672, 5)

Export Clean Data

In [94]:
# Exporting new dataframe as products_checked
df_prods.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# Step 2 - Run the df.describe() function

In [96]:
# Run the df.describe() function on your df_ords dataframe:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

I noticed a few things:
1) days_since_prior_order:  25% & 50% values are greater than the max
2) order_number: The max seems incorrect.  It’s only 1 which means that everyone only placed 1 order.  It seems like there would be some repeat customers.  My theory is also supported by the fact that the 25% & 75% numbers are greater then the max


# Step 3 - Check for mixed-type data

In [97]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [99]:
# checking for mixed-type data 
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There is no mixed-type data 

# Step 4 - Check for mixed-type data

There was no mixed-type data found.

# Step 5 - Check for missing values

In [100]:
#checking for missing values in df_ords data frame
df_ords.isnull().sum() 

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [101]:
# create a dataframe for the missing values
df_ords_nan = df_ords [df_ords ['days_since_prior_order'].isnull () == True]

In [102]:
df_ords_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [104]:
# searching the number_of_orders column for missing values for each value in our column.  
df_ords['order_number'].value_counts(dropna = False)

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

At first glance it seems that the missing values are tied to order_number being 1.  Which makes sense.  If it’s a customer’s first order, there wouldn’t be a prior order to reference.

I also did a second check to count the number of times order_number equaled 1.  The result matched the number of missing values: 206209

Which confirmed my theory.  


# Step 6 - Address the missing values

The missing values in this case are relevant to our analysis helps to tell the story.  Therefore, I’ve decided to take no action. 

# Step 7 - Check for duplicates

In [105]:
# checking for duplicate values in df_ords data frame
df_ords_dups = df_ords[df_ords.duplicated()]

In [106]:
# viewing duplicates values for df_ords
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found

# Step 8 - Address the duplicates

In [ ]:
No duplicates found

# Step 9 - Export your final, cleaned df_prods and df_ords data as “.csv”

In [107]:
# Export df_ords as orders_checked
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))

In [108]:
# Export df_prods as products_checked
df_prods.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))